From: Ray Strode <rstrode@redhat.com>
Date: Thu, 26 Mar 2020 14:11:50 -0400
Subject: main: switch log file when switching mode

plymouthd can be run in various modes, for, e.g., boot up,
shutdown, and software upgrades.

The mode plymouthd is using can be changed at runtime.

The "boot" mode keeps a log of the console messages that
happen during boot up.  At the moment, when changing from
the "boot" mode to any other mode, the log file is kept
open.

That open file can cause problems during shutdown.

This commit makes sure the log file is properly closed when
the mode is changed from boot to another mode.

https://gitlab.freedesktop.org/plymouth/plymouth/issues/88
---
 src/main.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/main.c b/src/main.c
index ddc1883..3d370d3 100644
--- a/src/main.c
+++ b/src/main.c
@@ -162,6 +162,7 @@ static void on_quit (state_t       *state,
                      ply_trigger_t *quit_trigger);
 static bool sh_is_init (state_t *state);
 static void cancel_pending_delayed_show (state_t *state);
+static void prepare_logging (state_t *state);
 
 static ply_boot_splash_mode_t
 get_splash_mode_from_mode (ply_mode_t mode)
@@ -237,6 +238,10 @@ on_change_mode (state_t    *state,
         else
                 return;
 
+        if (state->session != NULL) {
+                prepare_logging (state);
+        }
+
         splash_mode = get_splash_mode_from_mode (state->mode);
 
         if (!ply_boot_splash_show (state->boot_splash, splash_mode)) {
@@ -795,6 +800,8 @@ prepare_logging (state_t *state)
                 return;
         }
 
+        ply_terminal_session_close_log (state->session);
+
         logfile = get_log_file_for_state (state);
         if (logfile != NULL) {
                 bool log_opened;
